src/index.js

```diff
@@ -197,172 +328,189 @@ function _renderToString(vnode, context, isSvgMode, selectValue, parent) {
+
+					switch (name) {
+						case 'children':
+							children = v;
+							continue;
+
+						// VDOM-specific props
+						case 'key':
+						case 'ref':
+						case '__self':
+						case '__source':
+							continue;
+
+						// prefer for/class over htmlFor/className
+						case 'htmlFor':
+							if ('for' in props) continue;
+							name = 'for';
+							break;
+						case 'className':
+							if ('class' in props) continue;
+							name = 'class';
+							break;
+
+						// Form element reflected properties
+						case 'defaultChecked':
+							name = 'checked';
+							break;
+						case 'defaultSelected':
+							name = 'selected';
+							break;
+
+						// Special value attribute handling
+						case 'defaultValue':
+						case 'value':
+							name = 'value';
+							switch (type) {
+								// <textarea value="a&b"> --> <textarea>a&amp;b</textarea>
+								case 'textarea':
+									children = v;
+									continue;
+
+								// <select value> is serialized as a selected attribute on the matching option child
+								case 'select':
+									selectValue = v;
+									continue;
+
+								// Add a selected attribute to <option> if its value matches the parent <select> value
+								case 'option':
+									if (selectValue == v && !('selected' in props)) {
+										s = s + ' selected';
+									}
+									break;
+							}
+							break;
+
+						case 'dangerouslySetInnerHTML':
+							html = v && v.__html;
+							continue;
+
+						// serialize object styles to a CSS string
+						case 'style':
+							if (typeof v === 'object') {
+								v = styleObjToCss(v);
+							}
+							break;
+
+						default: {
+							if (isSvgMode && XLINK.test(name)) {
+								name = name
+									.toLowerCase()
+									.replace(XLINK_REPLACE_REGEX, 'xlink:');
+							} else if (UNSAFE_NAME.test(name)) {
+								continue;
+							} else if (name[0] === 'a' && name[1] === 'r' && v != null) {
+								// serialize boolean aria-xyz attribute values as strings
+								v += '';
+							}
+						}
+					}
+
+					// write this attribute to the buffer
+					if (v != null && v !== false && typeof v !== 'function') {
+						if (v === true || v === '') {
+							s = s + ' ' + name;
+						} else {
+							s = s + ' ' + name + '="' + encodeEntities(v + '') + '"';
+						}
+					}
 				}
-				break;
 
-			default: {
-				if (isSvgMode && XLINK.test(name)) {
-					name = name.toLowerCase().replace(XLINK_REPLACE_REGEX, 'xlink:');
-				} else if (UNSAFE_NAME.test(name)) {
-					continue;
-				} else if (name[0] === 'a' && name[1] === 'r' && v != null) {
-					// serialize boolean aria-xyz attribute values as strings
-					v += '';
+				if (UNSAFE_NAME.test(type)) {
+					throw new Error(`${type} is not a valid HTML tag name in ${s}>`);
+				} else if (html) {
+					// dangerouslySetInnerHTML defined this node's contents
+					output += s + '>' + html + '</' + type + '>';
+				} else if (children != null && typeof children !== 'boolean') {
```

Comment:
This removed an optimization we had in the normal iterator where we make single-text children a hot-path by already handling them, we could try that here as well to see the impact